### PR TITLE
WebAssembly 2022 queries

### PR DESCRIPTION
**Pull request by JamieWhitMac**
Queries for the WebAssembly 2022 chapter ([#2885](https://github.com/HTTPArchive/almanac.httparchive.org/issues/2885))

## WebAssembly adoption
- [x] % of desktop and mobile websites that use wasm
- [x] Ranks of pages which use wasm

## Popular WebAssembly languages
- [x] counts of language usage across wasm modules

## How often do we re-use WebAssembly libraries?
- [x] counts of all and unique wasm modules requested
- [x] count of wasm modules requested cross-origin

## Popular WebAssembly-powered Libraries
- [x] usages of popular wasm modules

## WASM module sizes
- [x] sizes of wasm module files and raw requests
- [x] size distribution of each section type

## Post-MVP Extensions
- [x] count of usages of post-MVP wasm extensions

### CODE REVIEW

**Comment by siakaramalegos** [comment]
Great job so far! Can you update the checkboxes so we get a better idea of how much is left to go?

**Comment by siakaramalegos** [comment]
Hi @JamieWhitMac it looks like you've written a lot of the queries, but I can't tell if they are finished yet. Can you update your progress using the checkboxes and/or commenting where you're at? Let me know if you need any help.

**Comment by JamieWhitMac** [comment]
Hi @siakaramalegos, apologies for the slow response. All but two queries are complete as the majority are re-used from last year's Almanac (aside from needing tweaks to dates to bring them up to date, so none are technically 'complete'). I'll be writing the remaining ones, changing the dates on the existing queries, and compiling the results in the coming week.

**Comment by siakaramalegos** [comment]
@JamieWhitMac just checking in - the analysis needs to be done by the end of the month (in just a few days) to stay on track for writing and publication. How is it going? Do you think you can still hit that deadline? Thanks!

**Comment by JamieWhitMac** [comment]
Hi @siakaramalegos, I encountered a little setback last week as I had a mistaken assumption about where the wasm data from the 2022 crawls would be. I've updated the comment above to show current progress changing the existing queries, hoping to have them finished ASAP.

**Comment by siakaramalegos** [comment]
@JamieWhitMac thanks for the update! Also, the analysts channel in Slack can be super helpful if you have a question or need help on a query.

**Comment by rviscomi** [comment]
@JamieWhitMac is this PR ready for review?

**Comment by JamieWhitMac** [comment]
@rviscomi it's very close, apologies for delay - should be ready to go in the next couple of days, once I double-check the linting and confirm the data we have is definitely everything we need.

**Comment by rviscomi** [comment]
Just noticed the linter wasn't running automatically because you're not in the HTTPArchive org on GitHub. I've manually kicked that off and sent you an invite to join now. (Which also gives you more privileges to edit stuff in this repo, like the checklist in #2885)

Also consider marking the PR ready for review now to get early feedback and we can merge it sooner. If there are any follow-up queries needed, those can be added in a separate PR.

File: sql/2022/webassembly/page_rankings.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# standardSQL
+# Count of pages which use wasm at page ranking intervals.
+
+SELECT
```
Comment (rviscomi) [inline sql/2022/webassembly/page_rankings.sql:4]: This query should also be grouping by client

File: sql/2022/webassembly/language_usage.sql [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    `httparchive.almanac.requests`
+  WHERE
+    date = '2022-06-01' AND
+    (mimeType = 'application/wasm' OR ext = 'wasm')
+)
+
+SELECT
+  client,
+  JSON_VALUE(wasm_stats, '$.language') AS language,
+  COUNT(0) AS count
```
Comment (rviscomi) [inline sql/2022/webassembly/language_usage.sql:18]: I'd also suggest calculating this as a %
```suggestion
  COUNT(0) AS count,
  SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
  COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
```

File: sql/2022/webassembly/popular_by_name.sql [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    ELSE
+      REGEXP_REPLACE(
+        REGEXP_EXTRACT(LOWER(url), r'.*/([^./?]*)'), -- lowercase & extract filename between last `/` and `.` or `?`
+        r'-[0-9a-f]{20,32}$', -- trim trailing hashes to transform `name-0abc43234[...]` to `name`
+        ''
+      )
+  END AS name,
+  COUNT(0) AS count,
+  COUNT(DISTINCT NET.REG_DOMAIN(url)) AS count_serving_hosts,
+  MIN(url) AS url
```
Comment (rviscomi) [inline sql/2022/webassembly/popular_by_name.sql:20]: Is the URL needed?
```suggestion
```

File: sql/2022/webassembly/proposals.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  FROM
+    `httparchive.almanac.requests`
+  WHERE
+    date = '2022-06-01' AND
+    (mimeType = 'application/wasm' OR ext = 'wasm')
+)
+
+SELECT
+  client,
+  COUNTIF(SAFE_CAST(JSON_VALUE(wasm_stats, '$.instr.proposals.atomics') AS INT64) > 0) AS atomics,
```
Comment (rviscomi) [inline sql/2022/webassembly/proposals.sql:17]: Do you also want these stats as %s?

**Review by rviscomi — CHANGES_REQUESTED** [review]

File: sql/2022/webassembly/page_rankings.sql [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  _rank AS rank,
```
Comment (rviscomi) [inline sql/2022/webassembly/page_rankings.sql:5]:
```suggestion
  client,
  _rank AS rank,
```

File: sql/2022/webassembly/page_rankings.sql [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  COUNT(DISTINCT page) AS pages
+FROM
+  `httparchive.almanac.requests`,
+  UNNEST([1000, 10000, 100000, 1000000, 10000000]) AS _rank
+WHERE
+  date = '2022-06-01' AND
+  (mimeType = 'application/wasm' OR ext = 'wasm') AND
+  rank <= _rank
+GROUP BY
+  rank
```
Comment (rviscomi) [inline sql/2022/webassembly/page_rankings.sql:15]:
```suggestion
  client,
  rank
```

**Review by rviscomi — APPROVED** [review]

**Comment by rviscomi** [comment]
Applied code review suggestions. I see that the results sheet seems to have the right data already, so the queries don't necessarily need to be rerun. Merging now.

**Pull request merged**